source/common/common/mutex_tracer.h

```diff
@@ -0,0 +1,54 @@
+#pragma once
+
+#include <atomic>
+#include <cstddef>
+#include <cstdint>
+
+#include "absl/base/internal/spinlock.h"
+#include "absl/base/thread_annotations.h"
+
+namespace Envoy {
+
+// Encapsulates a contention hook which can be registered with absl::RegisterMutexTracer() which
+// records statistics about that contention.
+//
+// MutexTracer should be accessed via getOrCreateTracer(), which ensures that the global singleton
+// MutexTracer object is always being called. This is necessary because of the type signature which
+// absl::RegisterMutexTracer() expects.
+//
+// This class is thread-safe, and can be called from multiple mutexes in contention across multiple
+// threads. This is made possible by utilizing memory_order_relaxed atomic writes.
+class MutexTracer final {
+public:
+  static MutexTracer* getOrCreateTracer();
+
+  // Returns the callback which can be registered via
+  // absl::RegisterMutexTracer(&Envoy::MutexTracer::contentionHook).
```

Comment:
I think this class should handle the registration, and thus contentionHook should be private.